### ISSUE TITLE
PCGUARD instrumentation may generate crashing binary

### ISSUE THREAD

**Issue by kcwu** [AFLplusplus/AFLplusplus#1120]
**Describe the bug**
The binary generated by CMPLOG may have different behavior.

**To Reproduce**
```
$ wget http://invisible-island.net/datafiles/release/lynx-cur.tar.gz
$ tar zxvf lynx-cur.tar.gz
$ cd lynx2.9.0dev.9
$ CC=afl-clang-lto CXX=afl-clang-lto++ ./configure --enable-debug
$ AFL_LLVM_CMPLOG=1 make clean all
```

The last step failed. To reproduce the exact command:
```
$ cd src/chrtrans
$ ./makeuctb ./hp_uni.tbl hp_uni.h
munmap_chunk(): invalid pointer
```

I found "makeuctb" crash only with CMPLOG+ -O2 or above optimization.

**Expected behavior**
`make` works

**Additional context**
I'm not sure it is a bug of CMPLOG or `makeuctb`. However, `makeuctb` without CMPLOG works well with asan and valgrind.

My clang version is 11.0.1-2.
Debian 5.10.46-4rodete4

**Comment by vanhauser-thc** [comment]
hmm works fine for me.
and the reported error has nothing to do with afl/lto.
could it be that you run out of memory/shmem?
try reboot and then `afl-system-config` ?

**Comment by kcwu** [comment]
With libdislocator.so, it crashed earlier. And I found the AFL map size is too small for some reason.

```
$ LD_PRELOAD=libdislocator.so gdb --args ./makeuctb ./hp_uni.tbl hp_uni.h
(gdb) watch __afl_area_ptr
Hardware watchpoint 1: __afl_area_ptr
(gdb) r
Starting program: /tmp/g/lynx2.9.0dev.9/src/chrtrans/makeuctb ./hp_uni.tbl hp_uni.h

Hardware watchpoint 1: __afl_area_ptr

Old value = (u8 *) 0x20f1f0 <__afl_area_initial> ""
New value = (u8 *) 0x7ffff7bdc000 '\314' <repeats 200 times>...
__afl_auto_first () at instrumentation/afl-compiler-rt.o.c:1323
c1323       __afl_area_ptr_backup = __afl_area_ptr;
(gdb) c
Continuing.

Hardware watchpoint 1: __afl_area_ptr

Old value = (u8 *) 0x7ffff7bdc000 '\314' <repeats 200 times>...
New value = (u8 *) 0x7ffff7fc2f08 '\314' <repeats 200 times>...
__afl_auto_second () at instrumentation/afl-compiler-rt.o.c:1301
1301          __afl_area_ptr_backup = __afl_area_ptr;
(gdb) p __afl_final_loc
$1 = 248
(gdb) c
Continuing.

Program received signal SIGSEGV, Segmentation fault.
0x000000000020c481 in addpair (fp=173, un=217) at ./makeuctb.c:238
238         if (un <= 0xfffe) {
(gdb) disassemble $pc-10,$pc+10
Dump of assembler code from 0x20c477 to 0x20c48b:
   0x000000000020c477 <addpair+1079>:   add    %al,(%rax)
   0x000000000020c479 <addpair+1081>:   add    %cl,-0x75(%rax)
   0x000000000020c47c <addpair+1084>:   add    $0x275f,%eax
=> 0x000000000020c481 <addpair+1089>:   mov    0xf8(%rax),%cl
   0x000000000020c487 <addpair+1095>:   add    $0x1,%cl
   0x000000000020c48a <addpair+1098>:   cmp    $0x1,%cl
End of assembler dump.
(gdb) p 0xf8
$2 = 248
```

**Comment by kcwu** [comment]
`AFL_DEBUG=1` output
```
[__afl_auto_init_globals]
DEBUG: debug enabled
DEBUG: (1) id_str <null>, __afl_area_ptr 0x7f30847f1f08, __afl_area_initial 0x20f1f0, __afl_area_ptr_dummy 0x20f1f0, __afl_map_addr 0x0, MAP_SIZE 65536, __afl_final_loc 248, max_size_forkserver 8388608/0x800000
DEBUG: (2) id_str <null>, __afl_area_ptr 0x7f30847f1f08, __afl_area_initial 0x20f1f0, __afl_area_ptr_dummy 0x20f1f0, __afl_map_addr 0x0, MAP_SIZE 65536, __afl_final_loc 248, max_size_forkserver 8388608/0x800000
DEBUG: cmplog id_str <null>
zsh: segmentation fault  AFL_DEBUG=1 LD_PRELOAD= ./makeuctb ./hp_uni.tbl hp_uni.h
```

**Comment by vanhauser-thc** [comment]
again, compiled with asan (`AFL_USE_ASAN=1`), works fine for me:

```
./makeuctb ./hp_uni.tbl hp_uni.h
[__afl_auto_init_globals]
DEBUG: debug enabled
DEBUG: (1) id_str <null>, __afl_area_ptr 0x616000000080, __afl_area_initial 0xbbf2e0, __afl_area_ptr_dummy 0xbbf2e0, __afl_map_addr 0x0, MAP_SIZE 65536, __afl_final_loc 616, max_size_forkserver 8388608/0x800000
DEBUG: (2) id_str <null>, __afl_area_ptr 0x616000000080, __afl_area_initial 0xbbf2e0, __afl_area_ptr_dummy 0xbbf2e0, __afl_map_addr 0x0, MAP_SIZE 65536, __afl_final_loc 616, max_size_forkserver 8388608/0x800000
DEBUG: cmplog id_str <null>
makeuctb: hp-roman8: translation map (_hp_roman8).
```

are you using the current dev branch state? have you done a "make clean install"?

**Comment by kcwu** [comment]
Yes I am using AFL++ the latest dev branch and have clean build.

I found this is related to llvm's version. 11.0.1 is bad and 13.0.0 is good. With `git bisect`, I found https://github.com/llvm/llvm-project/commit/9d172c8e9c845a36b61dc12c27de8acdbef8b247 is the commit changed the behavior. 

BTW, I found clang crashing with assertion failure when compile `makeuctb` before https://github.com/llvm/llvm-project/commit/296314516d103f8eeb987a08b509c1381cfeef89. 
<code>
clang: /llvm-project/llvm/include/llvm/ADT/PointerIntPair.h:178: static intptr_t llvm::PointerIntPairInfo<PointerT, IntBits, PtrTraits>::updatePointer(intptr_t, PointerT) [with PointerT = const llvm::MemoryBuffer*; unsigned int IntBits = 2; PtrTraits = llvm::PointerLikeTypeTraits<const llvm::MemoryBuffer*>; intptr_t = long int]: Assertion `(PtrWord & ~PointerBitMask) == 0 && "Pointer is not sufficiently aligned"' failed. 
</code>

So, if you want to build clang 11.x, `-DLLVM_ENABLE_ASSERTIONS=No` is needed.

In other words, only llvm 12.0.0 or above can avoid these two issues.

**Comment by vanhauser-thc** [comment]
During afl++ development we also found llvm bugs :)
kinda relieved its a llvm issue not an afl++ issue.
as I have never encountered this with any llvm version before it seems to be a corner case.

**Comment by kcwu** [comment]
I found another binary crashed with clang 12.0.1 + CMPLOG. And the symptom is similar (small `__afl_final_loc`). I will investigate and give more detail later.

**Comment by vanhauser-thc** [comment]
yes please!

**Comment by kcwu** [comment]
How to reproduce this issue

1. source code h.c
```c
int main(int argc, char *argv[]) {
  if (argc > 0) argc++;
  if (argc > 0) argc++;
  if (argc > 0) argc++;
  if (argc > 0) argc++;
  return 0;
}
```
2. compile
```
$ AFL_LLVM_INSTRUMENT=PCGUARD afl-clang-lto -g -O0 -o h h.c
afl-cc ++3.15a by Michal Zalewski, Laszlo Szekeres, Marc Heuse - mode: LLVM-LTO-PCGUARD
afl-llvm-lto++3.15a by Marc "vanHauser" Heuse <mh@mh-sec.de>
[+] Instrumented 8 locations with no collisions (on average 0 collisions would be in afl-gcc/vanilla AFL) (non-hardened mode).
```
3. note that `Instrumented 8 locations`. If the number is not 8, please modify the source code to make it multiple of 8 by adding or deleting `if` lines.
4. run with libdislocator
```
$ LD_PRELOAD=/path/to/AFLplusplus/utils/libdislocator/libdislocator.so  ./h
Segmentation fault
```

OTOH, `AFL_LLVM_INSTRUMENT=CLASSIC` works well.

I guess the problem is, other instrumentation, like [afl-llvm-lto-instrumentation](https://github.com/AFLplusplus/AFLplusplus/blob/stable/instrumentation/afl-llvm-lto-instrumentation.so.cc#L859) use `afl_global_id++` as id. But [SanitizerCoverageLTO](https://github.com/AFLplusplus/AFLplusplus/blob/stable/instrumentation/SanitizerCoverageLTO.so.cc#L1304) use `++afl_global_id` as id.
In other words, the last used id is `afl_global_id` and the map size need at least `afl_global_id+1` for `SanitizerCoverageLTO`. Unfortunately, https://github.com/AFLplusplus/AFLplusplus/blob/stable/instrumentation/SanitizerCoverageLTO.so.cc#L1057 won't plus 1 for multiple of 8.

https://github.com/AFLplusplus/AFLplusplus/pull/1135 is my proposed fix.
